ergo-lib/src/wallet/box_selector/simple.rs

```diff
@@ -152,29 +166,144 @@ impl<T: ErgoBoxAssets> BoxSelector<T> for SimpleBoxSelector {
+            format!("total value of the selected boxes {:?} should equal target balance {:?} + total value in change boxes {:?}", sum_selected_inputs, target_balance.as_u64(), sum_change_boxes)
+        ));
+    }
+
+    let sum_tokens_selected_inputs = sum_tokens_from_boxes(selected_inputs)?;
+    let sum_tokens_change_boxes = sum_tokens_from_boxes(change_boxes)?;
+    let sum_tokens_target = sum_tokens(Some(target_tokens))?;
+    if sum_tokens_selected_inputs
+        != sum_tokens_from_hashmaps(sum_tokens_change_boxes.clone(), sum_tokens_target.clone())?
+    {
+        return Err(CheckPreservationError(
+            format!("all tokens from selected boxes {:?} should equal all tokens from the change boxes {:?} + target tokens {:?}", sum_tokens_selected_inputs, sum_tokens_change_boxes, sum_tokens_target)
+        ));
+    }
+    Ok(())
+}
+
+/// Not enough coins for change box(es)
+#[derive(Error, PartialEq, Eq, Debug, Clone)]
+#[error("Not enough coins for change box(es)")]
+pub struct NotEnoughCoinsForChangeBox(String);
+
+/// Split change tokens into a multiple boxes if over 255 distinct tokens
+fn make_change_boxes(
+    change_value: BoxValue,
+    change_tokens: HashMap<TokenId, TokenAmount>,
+) -> Result<Vec<ErgoBoxAssetsData>, NotEnoughCoinsForChangeBox> {
+    if change_tokens.is_empty() {
+        Ok(vec![ErgoBoxAssetsData {
+            value: change_value,
+            tokens: None,
+        }])
+    } else if change_tokens.len() < ErgoBox::MAX_TOKENS_COUNT {
+        #[allow(clippy::unwrap_used)]
+        // unwrap_used is ok here because we checked that change_tokens.len() < ErgoBox::MAX_TOKENS_COUNT
+        Ok(vec![ErgoBoxAssetsData {
+            value: change_value,
+            tokens: Some(
+                BoxTokens::from_vec(change_tokens.into_iter().map(Token::from).collect()).unwrap(),
+            ),
+        }])
+    } else {
+        let mut change_boxes = vec![];
+        let mut change_tokens_left: Vec<Token> =
+            change_tokens.into_iter().map(Token::from).collect();
+        let mut change_value_left = change_value;
+        while !change_tokens_left.is_empty() {
+            if change_tokens_left.len() < ErgoBox::MAX_TOKENS_COUNT {
+                #[allow(clippy::unwrap_used)]
+                // unwrap_used is ok here because we checked that change_tokens_left.len() < ErgoBox::MAX_TOKENS_COUNT
+                let change_box = ErgoBoxAssetsData {
+                    value: change_value_left,
+                    tokens: Some(BoxTokens::from_vec(change_tokens_left).unwrap()),
+                };
+                change_boxes.push(change_box);
+                break;
+            } else {
+                #[allow(clippy::unwrap_used)] // safe for the box value upper bound
+                // doubled due to larger box size to accomodate so many tokens
+                let value = BoxValue::SAFE_USER_MIN.checked_mul_u32(2).unwrap();
+                let tokens_to_drain = ErgoBox::MAX_TOKENS_COUNT / 2;
```

Comment:
Small nitpick, but I think a more efficient way to distribute boxes would be to drain change_tokens.len() % (ErgoBox::MAX_TOKENS_COUNT + 1) each time. Right now if a user has say, 510 Tokens, the distribution will be done like so if I've traced the code correctly:

First Box has 127 tokens (MAX_TOKENS_COUNT / 2), 383 still remaining
Second Box has 127 tokens, 256 remaining
Third Box will have 127 tokens, 129 remaining
Fourth one will have 129

Whereas this could be done with two boxes, each having 255 tokens.